feat(genesis export): genesis exporting for assets, delegation and operator modules.

TimmyExogenous opened the pull request:
# Description

It provides the genesis exporting functions for the assets, delegation, and operator modules. It provides compatibility for starting the Exocore node from the bootStrap contract and general genesis file.  

Please refer to  this documentation to get more details:
[understanding-of-the-Genesis-exporting](https://www.notion.so/exocore/understanding-of-the-Genesis-exporting-b8d98d0535aa473a9d9ed4091749c027)

The main changes are the proto files related to the genesis state. All other changes are implemented around the redefined genesis state structure to enable the genesis import and export functionality.

* Assets module:  add `operator_assets` in the genesis state
* Operator module: add `OperatorConsKeyRecord,OptedState,AVSUSDValue,OperatorUSDValue,OperatorSlashState,PrevConsKey,OperatorKeyRemoval` in the genesis state. 
* delegation module: remove `delegations`, and add `delegation_states,stakers_by_operator,undelegations` in the genesis state. 

The purpose of these changes is to ensure the unbalanced states won't influence the exported state.

# Todo
* unit test
* integration test

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Enhanced initialization options with improved parameters for operator management.
  - Introduced new message structures for managing operator details and delegation states, improving data organization.
  - Updated asset management functionality with refined handling of operator-specific assets.
  - Added new fields in genesis state for comprehensive representation of delegation and operator information.
  - Improved data retrieval methods by changing return types from maps to slices for better performance.

- **Bug Fixes**
  - Corrected issues with struct field assignments in genesis state initialization to ensure accuracy.

- **Refactor**
  - Renamed instances of `ExocoreApp` to `exocoreApp` for clarity and consistency in application creation and export functions.
  - Updated function signatures and return types in various modules to improve data handling and clarity.

- **Tests**
  - Improved test coverage and assertion accuracy for genesis state validation and asset management functions.

- **Chores**
  - Updated the `git` package version in the Dockerfile for potential bug fixes and improvements.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TimmyExogenous committed:
refine the genesis validation and add implement the genesis exporting in assets module

TimmyExogenous committed:
add genesis exporting for the delegation module

TimmyExogenous committed:
move the isGeneralInit flag to the commandline of starting node

TimmyExogenous committed:
add genesis exporting for operator module

TimmyExogenous committed:
finish genesis exporting for operator module

TimmyExogenous committed:
Remove the isGeneralInit flag that indicates whether the launch is from bootstrap or the exported genesis file.
fix the issues caused by rebase

TimmyExogenous committed:
remove the opt in when init dogfood genesis
fix the issue from code comments
fix the error after rebasing

MaxMustermann2 commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -117,7 +117,8 @@ message StakerAllAssetsInfo {
 message OperatorAssetInfo {
-  // total_amount is the total amount of the asset deposited.
+  // total_amount is the total amount of the asset deposited, which excludes the
+  // wait_unbonding_amount
```
Why exclude that `wait_unbonding_amount` ? Is it because this is only supposed to represent the active state ?

TimmyExogenous replied:
Right,  it represents all active amounts of the asset pool, it will be used to calculate the delegated amount of a specified staker from its share. I will add a code comment to clarify this.

MaxMustermann2 commented on proto/exocore/operator/v1/genesis.proto:
```diff
@@ -12,10 +12,129 @@ option go_package = "github.com/ExocoreNetwork/exocore/x/operator/types";
+  // operator_records refers to a list of operator records. each record
+  // contains an operator address and a list of chain id +
+  // cons key combination.
+  repeated OperatorConsKeyRecord operator_records = 2
```
I had moved this field to `x/dogfood` which opts the operator (which must have been registered already) into the AVS. Do you think it's better located here?

TimmyExogenous replied:
I plan to remove the opt-in from the InitGenesis of Dogfood and retain the changes from #PR108 that set the consensus key and opted information in the operator module. This will ensure that the imported states remain consistent with the pre-export states—for example, the opted height will be preserved, and the USD value of operators won’t be reset to zero by the opt-in during genesis initialization. This needs to be tested to see if it works well.

TimmyExogenous committed:
detail the return error for operator info

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous commented on x/operator/keeper/slash.go:
```diff
@@ -43,9 +41,12 @@ func SlashFromUndelegation(undelegation *delegationtype.UndelegationRecord, slas
 }
 
 func (k *Keeper) CheckSlashParameter(ctx sdk.Context, parameter *types.SlashInputInfo) error {
+	if parameter.SlashProportion.IsNil() || parameter.SlashProportion.IsNegative() {
+		return errorsmod.Wrapf(types.ErrValueIsNilOrZero, "SlashProportion: %+v", parameter.SlashProportion)
+	}
```
thanks

TimmyExogenous commented on x/assets/types/genesis.go:
```diff
@@ -29,9 +36,337 @@
+	return seen, nil
+}
+
+// ValidateClientChains performs basic client chains validation
+func (gs GenesisState) ValidateClientChains() (map[uint64]struct{}, error) {
+	validationFunc := func(i int, info ClientChainInfo) error {
+		if info.Name == "" {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"nil Name for chain %d",
+				i,
+			)
+		}
+		// this is our primary method of cross-chain communication.
+		if info.LayerZeroChainID == 0 {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"nil LayerZeroChainID for chain %s",
+				info.Name,
+			)
+		}
+		// the address length is used to convert from bytes32 to address.
+		if info.AddressLength == 0 {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"nil AddressLength for chain %s",
+				info.Name,
+			)
+		}
+		return nil
+	}
+	seenFieldValueFunc := func(info ClientChainInfo) (uint64, struct{}) {
+		return info.LayerZeroChainID, struct{}{}
+	}
+	lzIDs, err := CommonValidation(gs.ClientChains, seenFieldValueFunc, validationFunc)
+	if err != nil {
+		return nil, err
+	}
+	return lzIDs, nil
+}
```
It will be used in the log print.

TimmyExogenous commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -228,6 +255,33 @@ func (k *Keeper) GetStakersByOperator(ctx sdk.Context, operator, assetID string)
+		})
+	}
+	return ret, nil
+}
```
follow the code style of the comos-sdk.

TimmyExogenous commented on cmd/exocored/testnet.go:
```diff
@@ -399,7 +399,7 @@ func getTestExocoreGenesis(
 	depositAmount := sdk.TokensFromConsensusPower(power, evmostypes.PowerReduction)
 	depositsByStaker := []assetstypes.DepositsByStaker{}
-	operatorInfos := []operatortypes.OperatorInfo{}
+	operatorInfos := []operatortypes.OperatorDetail{}
```
These codes aren't used, and they are from the other Repo, so ignore this issue.

TimmyExogenous commented on x/slash/keeper/execute_slash_test.go:
```diff
@@ -66,5 +66,5 @@ func (suite *SlashTestSuite) TestSlash() {
 	assetInfo, err := suite.App.AssetsKeeper.GetStakingAssetInfo(suite.Ctx, assetID)
 	suite.NoError(err)
-	suite.Equal(assets[assetID].StakingTotalAmount.Add(depositEvent.OpAmount).Sub(event.OpAmount), assetInfo.StakingTotalAmount)
+	suite.Equal(assets[0].StakingTotalAmount.Add(depositEvent.OpAmount).Sub(event.OpAmount), assetInfo.StakingTotalAmount)
```
This module hasn't been used, the related code might be removed. So ignore this issue.

TimmyExogenous commented on x/delegation/keeper/un_delegation_state.go:
```diff
@@ -13,17 +13,34 @@ import (
+func (k *Keeper) SetUndelegationRecords(ctx sdk.Context, records []types.UndelegationRecord) error {
 	singleRecordStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixUndelegationInfo)
 	stakerUndelegationStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixStakerUndelegationInfo)
 	pendingUndelegationStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixPendingUndelegations)
-	// key := common.HexToAddress(incentive.Contract)
-	for _, record := range records {
-		bz := k.cdc.MustMarshal(record)
+	currentHeight := ctx.BlockHeight()
+	for i := range records {
+		record := records[i]
+		if record.CompleteBlockNumber < uint64(currentHeight) {
+			return errorsmod.Wrapf(types.ErrInvalidCompletedHeight, "currentHeight:%d,CompleteBlockNumber:%d", currentHeight, record.CompleteBlockNumber)
+		}
```
The completion will be executed at the end of the block, so the completed block number might be equal to the current height.

TimmyExogenous commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -36,9 +37,9 @@
 message QueryAllStakingAssetsInfoResponse {
   // all_staking_assets_info is the response for all staking assets info.
-  map<string, StakingAssetInfo> all_staking_assets_info = 1;
+  repeated StakingAssetInfo all_staking_assets_info = 1 [(gogoproto.nullable) = false];
```
The existing clients have updated this RPC.

TimmyExogenous commented on proto/exocore/operator/v1/genesis.proto:
```diff
@@ -12,15 +12,88 @@
   // there are no params for this module.
   // operators is a list of the registered operators.
-  repeated OperatorInfo operators = 1 [(gogoproto.nullable) = false];
+  repeated OperatorDetail operators = 1 [(gogoproto.nullable) = false];
```
Thanks for the reminder. The affection has been considered and addressed.

TimmyExogenous replied:
we will use Bootstrap to launch the next version of Testnet. So the data will be reset.

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -21,15 +21,8 @@
-	}
 	// #nosec G703 // already validated in `ValidateBasic`
-	opAccAddr, _ := sdk.AccAddressFromBech32(info.EarningsAddr)
+	opAccAddr, _ := sdk.AccAddressFromBech32(addr)
```
I will fix it in the next commit.

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -79,16 +72,21 @@
+func (k *Keeper) AllOperators(ctx sdk.Context) []operatortypes.OperatorDetail {
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), operatortypes.KeyPrefixOperatorInfo)
 	iterator := sdk.KVStorePrefixIterator(store, nil)
 	defer iterator.Close()
 
-	ret := make([]string, 0)
+	ret := make([]operatortypes.OperatorDetail, 0)
 	for ; iterator.Valid(); iterator.Next() {
-		accAddr := sdk.AccAddress(iterator.Key())
-		ret = append(ret, accAddr.String())
+		var operatorInfo operatortypes.OperatorInfo
+		operatorAddr := sdk.AccAddress(iterator.Key())
+		k.cdc.MustUnmarshal(iterator.Value(), &operatorInfo)
+		ret = append(ret, operatortypes.OperatorDetail{
+			OperatorAddress: operatorAddr.String(),
+			OperatorInfo:    operatorInfo,
+		})
```
The number of operators won't be too large, so pagination is not considered for now.

TimmyExogenous commented on x/operator/types/expected_keepers.go:
```diff
@@ -27,7 +27,7 @@ type AssetsKeeper interface {
 	) error
 	ClientChainExists(ctx sdk.Context, index uint64) bool
-	GetAllStakingAssetsInfo(ctx sdk.Context) (allAssets map[string]*assetstype.StakingAssetInfo, err error)
+	GetAllStakingAssetsInfo(ctx sdk.Context) (allAssets []assetstype.StakingAssetInfo, err error)
```
This method isn't expected by the above files, and all codes related to this method has been updated.

TimmyExogenous commented on proto/exocore/assets/v1/genesis.proto:
```diff
@@ -27,13 +27,38 @@ message GenesisState {
   // which contains deposits, withdrawable and unbonding amount.
   // at genesis (not chain restart), the unbonding amount must be 0.
   repeated DepositsByStaker deposits = 4 [(gogoproto.nullable) = false];
+
+  // operator_assets is the list of all operator assets information, indexed
+  // by the operator address and the asset id. The struct is the `OperatorAssetInfo`
+  repeated AssetsByOperator operator_assets = 5 [(gogoproto.nullable) = false];
+}
```
These have been implemented in the `InitGenesis` and `ExportGenesis`

TimmyExogenous commented on x/delegation/keeper/genesis.go:
```diff
@@ -50,12 +78,44 @@
+	err = k.SetUndelegationRecords(ctx, gs.Undelegations)
+	if err != nil {
+		panic(err)
+	}
```
These three methods are low-level functions used to batch set related states to the KVStore. The code style follows the Cosmos-SDK. All inputs are checked before these methods are called, so this suggestion doesn't seem too necessary.

TimmyExogenous commented on x/assets/keeper/client_chain_asset.go:
```diff
@@ -88,17 +88,16 @@ func (k Keeper) GetAssetsDecimal(ctx sdk.Context, assets map[string]interface{})
+func (k Keeper) GetAllStakingAssetsInfo(ctx sdk.Context) (allAssets []assetstype.StakingAssetInfo, err error) {
 	store := ctx.KVStore(k.storeKey)
 	iterator := sdk.KVStorePrefixIterator(store, assetstype.KeyPrefixReStakingAssetInfo)
 	defer iterator.Close()
 
-	ret := make(map[string]*assetstype.StakingAssetInfo, 0)
+	ret := make([]assetstype.StakingAssetInfo, 0)
 	for ; iterator.Valid(); iterator.Next() {
 		var assetInfo assetstype.StakingAssetInfo
 		k.cdc.MustUnmarshal(iterator.Value(), &assetInfo)
-		_, assetID := assetstype.GetStakeIDAndAssetIDFromStr(assetInfo.AssetBasicInfo.LayerZeroChainID, "", assetInfo.AssetBasicInfo.Address)
-		ret[assetID] = &assetInfo
+		ret = append(ret, assetInfo)
```
The implementation follows the code style of the Cosmos-SDK. The node won't function properly if the data is corrupted, so this case won't be addressed.

TimmyExogenous commented on x/delegation/keeper/un_delegation_state.go:
```diff
@@ -13,17 +13,34 @@ import (
 	"github.com/ethereum/go-ethereum/common/hexutil"
 )
 
+func (k Keeper) AllUndelegations(ctx sdk.Context) (undelegations []types.UndelegationRecord, err error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixUndelegationInfo)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]types.UndelegationRecord, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		var undelegation types.UndelegationRecord
+		k.cdc.MustUnmarshal(iterator.Value(), &undelegation)
+		ret = append(ret, undelegation)
+	}
+	return ret, nil
```
The implementation follows the code style of the Cosmos-SDK. The node won't function properly if the data is corrupted, so this case won't be addressed.

TimmyExogenous commented on x/delegation/keeper/genesis.go:
```diff
@@ -66,12 +31,48 @@ func (k Keeper) InitGenesis(
 			panic(errorsmod.Wrap(err, "failed to associate operator with staker"))
 		}
 	}
+
+	// init the state from the general exporting genesis file
+	err := k.SetAllDelegationStates(ctx, gs.DelegationStates)
+	if err != nil {
+		panic(err)
+	}
+	err = k.SetAllStakerList(ctx, gs.StakersByOperator)
+	if err != nil {
+		panic(err)
+	}
+	err = k.SetUndelegationRecords(ctx, gs.Undelegations)
+	if err != nil {
+		panic(err)
+	}
```
The panic will print the error info, I will detail the error description.

TimmyExogenous commented on x/operator/keeper/genesis.go:
```diff
@@ -9,14 +9,96 @@ import (
+		}
+		if err := k.SetOperatorInfo(ctx, op.OperatorAddress, &op.OperatorInfo); err != nil {
 			panic(err)
 		}
 	}
+	for _, record := range state.OperatorRecords {
+		addr := record.OperatorAddress
+		// #nosec G703 // already validated
+		operatorAddr, _ := sdk.AccAddressFromBech32(addr)
+		for _, detail := range record.Chains {
+			wrappedKey := types.NewWrappedConsKeyFromHex(detail.ConsensusKey)
+			bz := k.cdc.MustMarshal(wrappedKey.ToTmProtoKey())
+			k.setOperatorConsKeyForChainIDUnchecked(ctx, operatorAddr, wrappedKey.ToConsAddr(), detail.ChainID, bz)
+		}
+	}
+	// init the state from the general exporting genesis file
+	err := k.SetAllOptedInfo(ctx, state.OptStates)
+	if err != nil {
+		panic(err)
+	}
+	err = k.SetAllOperatorUSDValues(ctx, state.OperatorUSDValues)
+	if err != nil {
+		panic(err)
+	}
+	err = k.SetAllAVSUSDValues(ctx, state.AVSUSDValues)
+	if err != nil {
+		panic(err)
+	}
+	err = k.SetAllSlashInfo(ctx, state.SlashStates)
+	if err != nil {
+		panic(err)
+	}
+	err = k.SetAllPrevConsKeys(ctx, state.PreConsKeys)
+	if err != nil {
+		panic(err)
+	}
+	err = k.SetAllOperatorKeyRemovals(ctx, state.OperatorKeyRemovals)
+	if err != nil {
+		panic(err)
+	}
```
The `InitGenesis` is an interface defined by Cosmos-SDK, it doesn't return an error. And the node should panic if there is an error when initializing the genesis file. I will detail the error info when panic.

TimmyExogenous commented on x/delegation/keeper/genesis.go:
```diff
@@ -66,12 +31,48 @@
-	genesis := delegationtype.DefaultGenesis()
-	// TODO
-	return genesis
+func (k Keeper) ExportGenesis(ctx sdk.Context) *delegationtype.GenesisState {
+	res := delegationtype.GenesisState{}
+	associations, err := k.GetAllAssociations(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.Associations = associations
+
+	delegationStates, err := k.AllDelegationStates(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.DelegationStates = delegationStates
+
+	stakerList, err := k.AllStakerList(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.StakersByOperator = stakerList
+
+	undelegations, err := k.AllUndelegations(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.Undelegations = undelegations
+	return &res
```
I will change the panic to print a detailed error, so it won't affect the node's operation if there is an error during genesis export.

TimmyExogenous commented on x/operator/keeper/genesis.go:
```diff
@@ -9,14 +9,96 @@
+	}
+	res.OperatorRecords = operatorRecords
+
+	optedInfos, err := k.GetAllOptedInfo(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.OptStates = optedInfos
+
+	allAVSUSDValues, err := k.GetAllAVSUSDValues(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.AVSUSDValues = allAVSUSDValues
+
+	allOperatorUSDValues, err := k.GetAllOperatorUSDValues(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.OperatorUSDValues = allOperatorUSDValues
+
+	slashingInfos, err := k.GetAllSlashInfo(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.SlashStates = slashingInfos
+
+	prevConsKeys, err := k.GetAllPrevConsKeys(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.PreConsKeys = prevConsKeys
+
+	operatorKeyRemovals, err := k.GetAllOperatorKeyRemovals(ctx)
+	if err != nil {
+		panic(err)
+	}
+	res.OperatorKeyRemovals = operatorKeyRemovals
+
+	return &res
```
I will change the panic to print a detailed error, so it won't affect the node's operation if there is an error during genesis export.

TimmyExogenous commented on x/operator/keeper/usd_value.go:
```diff
@@ -240,3 +240,57 @@ func (k *Keeper) GetOperatorAssetValue(ctx sdk.Context, operator sdk.AccAddress,
 	// truncate the USD value to int64
 	return optedUSDValues.ActiveUSDValue.TruncateInt64(), nil
 }
+
+func (k *Keeper) SetAllOperatorUSDValues(ctx sdk.Context, usdValues []operatortypes.OperatorUSDValue) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), operatortypes.KeyPrefixUSDValueForOperator)
+	for i := range usdValues {
+		usdValue := usdValues[i]
+		bz := k.cdc.MustMarshal(&usdValue)
+		store.Set([]byte(usdValue.Key), bz)
+	}
+	return nil
+}
```
The Cosmos KVStore doesn't provide a function to set a batch of states.

TimmyExogenous commented on x/operator/keeper/usd_value.go:
```diff
@@ -240,3 +240,57 @@
+	return ret, nil
+}
+
+func (k *Keeper) SetAllAVSUSDValues(ctx sdk.Context, usdValues []operatortypes.AVSUSDValue) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), operatortypes.KeyPrefixUSDValueForAVS)
+	for i := range usdValues {
+		usdValue := usdValues[i]
+		bz := k.cdc.MustMarshal(&usdValue.Value)
+		store.Set([]byte(usdValue.AVSAddr), bz)
+	}
+	return nil
+}
```
The Cosmos KVStore doesn't provide a function to set a batch of states.

TimmyExogenous commented on x/operator/types/keys.go:
```diff
@@ -126,13 +132,64 @@ func KeyForOperatorAndChainIDToConsKey(addr sdk.AccAddress, chainID string) []by
 	)
 }
 
+func ParseKeyForOperatorAndChainIDToConsKey(key []byte) (addr sdk.AccAddress, chainID string, err error) {
+	// Extract the address
+	addr = key[0:AccAddressLength]
+	if len(addr) == 0 {
+		return nil, "", xerrors.New("missing address")
+	}
+
+	// Extract the chainID length
+	chainIDLen := sdk.BigEndianToUint64(key[AccAddressLength : AccAddressLength+8])
+	if len(key) != int(AccAddressLength+8+chainIDLen) {
+		return nil, "", xerrors.Errorf("invalid key length,expected:%d,got:%d", AccAddressLength+8+chainIDLen, len(key))
+	}
+
+	// Extract the chainID
+	chainIDBytes := key[AccAddressLength+8 : AccAddressLength+chainIDLen]
+	chainID = string(chainIDBytes)
+
+	return addr, chainID, nil
+}
```
Thanks, I will add the check in the next commit.

TimmyExogenous commented on x/operator/types/keys.go:
```diff
@@ -126,13 +132,64 @@
 	)
 }
 
+func ParsePrevConsKey(key []byte) (chainID string, addr sdk.AccAddress, err error) {
+	// Check if the key has at least one byte for the prefix and one byte for the chainID length
+	if len(key) < 2 {
+		return "", nil, xerrors.New("key too short")
+	}
+
+	// Extract and validate the prefix
+	prefix := key[0]
+	if prefix != BytePrefixForOperatorAndChainIDToPrevConsKey {
+		return "", nil, xerrors.Errorf("invalid prefix: expected %x, got %x", BytePrefixForOperatorAndChainIDToPrevConsKey, prefix)
+	}
+
+	// Extract the chainID length
+	chainIDLen := sdk.BigEndianToUint64(key[1:9])
+	if len(key) < int(9+chainIDLen) {
+		return "", nil, xerrors.New("key too short for chainID length")
+	}
+
+	// Extract the chainID
+	chainIDBytes := key[9 : 9+chainIDLen]
+	chainID = string(chainIDBytes)
+
+	// Extract the address
+	addr = key[9+chainIDLen:]
+	if len(addr) == 0 {
+		return "", nil, xerrors.New("missing address")
+	}
+
+	return chainID, addr, nil
+}
```
I will fix it in the next commit

TimmyExogenous commented on x/operator/types/keys.go:
```diff
@@ -155,6 +212,37 @@
 	)
 }
 
+func ParseKeyForOperatorKeyRemoval(key []byte) (addr sdk.AccAddress, chainID string, err error) {
+	// Check if the key has at least one byte for the prefix and one byte for the chainID length
+	if len(key) < 2 {
+		return nil, "", xerrors.New("key too short")
+	}
+
+	// Extract and validate the prefix
+	prefix := key[0]
+	if prefix != BytePrefixForOperatorKeyRemovalForChainID {
+		return nil, "", xerrors.Errorf("invalid prefix: expected %x, got %x", BytePrefixForOperatorKeyRemovalForChainID, prefix)
+	}
+
+	// Extract the address
+	addr = key[1 : AccAddressLength+1]
+	if len(addr) == 0 {
+		return nil, "", xerrors.New("missing address")
+	}
+
+	// Extract the chainID length
+	chainIDLen := sdk.BigEndianToUint64(key[AccAddressLength+1 : AccAddressLength+9])
+	if len(key) != int(AccAddressLength+9+chainIDLen) {
+		return nil, "", xerrors.Errorf("invalid key length,expected:%d,got:%d", AccAddressLength+9+chainIDLen, len(key))
+	}
+
+	// Extract the chainID
+	chainIDBytes := key[AccAddressLength+9 : AccAddressLength+9+chainIDLen]
+	chainID = string(chainIDBytes)
+
+	return addr, chainID, nil
+}
```
It will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/consensus_keys.go:
```diff
@@ -485,3 +488,93 @@
+		store.Set(keyBytes, []byte{})
+	}
+	return nil
+}
```
It will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/consensus_keys.go:
```diff
@@ -485,3 +488,93 @@
+		store.Set(keyBytes, bz)
+	}
+	return nil
+}
```
it will be fixed in the next commit

TimmyExogenous commented on x/operator/client/cli/tx.go:
```diff
@@ -112,11 +115,16 @@ func newBuildRegisterOperatorMsg(
 	if approveAddr == "" {
 		approveAddr = sender.String()
 	}
+	// #nosec G703 // this only errors if the flag isn't defined.
+	earningAddr, _ := fs.GetString(FlagEarningAddr)
+	if earningAddr == "" {
+		earningAddr = sender.String()
+	}
```
this only errors if the flag isn't defined, so ignore this issue.

TimmyExogenous commented on x/operator/keeper/operator_slash_state.go:
```diff
@@ -155,3 +155,30 @@ func (k *Keeper) GetSlashAssetsState(ctx sdk.Context, assetID, stakerOrOperator
 	return ret.Amount, nil
 }
+
+func (k *Keeper) SetAllSlashInfo(ctx sdk.Context, slashInfos []operatortypes.OperatorSlashState) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), operatortypes.KeyPrefixOperatorSlashInfo)
+	for i := range slashInfos {
+		info := slashInfos[i]
+		bz := k.cdc.MustMarshal(&info)
+		store.Set([]byte(info.Key), bz)
+	}
+	return nil
+}
```
This function is only used in genesis initiation, and any error will cause a panic. So this reminder is unnecessary.

TimmyExogenous commented on x/operator/keeper/operator_slash_state.go:
```diff
@@ -155,3 +155,30 @@
+	return nil
+}
+
+func (k *Keeper) GetAllSlashInfo(ctx sdk.Context) ([]operatortypes.OperatorSlashState, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), operatortypes.KeyPrefixOperatorSlashInfo)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]operatortypes.OperatorSlashState, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		var slashInfo operatortypes.OperatorSlashInfo
+		k.cdc.MustUnmarshal(iterator.Value(), &slashInfo)
+		ret = append(ret, operatortypes.OperatorSlashState{
+			Key:  string(iterator.Key()),
+			Info: slashInfo,
+		})
+	}
+	return ret, nil
+}
```
There won't be too many slash records, so we don't consider this now.

TimmyExogenous commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -14,6 +14,33 @@ import (
+	return ret, nil
+}
+
+func (k Keeper) SetAllDelegationStates(ctx sdk.Context, delegationStates []delegationtype.DelegationStates) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), delegationtype.KeyPrefixRestakerDelegationInfo)
+	for i := range delegationStates {
+		singleElement := delegationStates[i]
+		bz := k.cdc.MustMarshal(&singleElement.States)
+		store.Set([]byte(singleElement.Key), bz)
+	}
+	return nil
+}
```
It follows the code style of Cosmos SDK, it's unnecessary to address the error here.

TimmyExogenous commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -249,6 +276,33 @@
+	return ret, nil
+}
+
+func (k Keeper) SetAllStakerList(ctx sdk.Context, stakersByOperator []delegationtype.StakersByOperator) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), delegationtype.KeyPrefixStakersByOperator)
+	for i := range stakersByOperator {
+		singleElement := stakersByOperator[i]
+		bz := k.cdc.MustMarshal(&delegationtype.StakerList{Stakers: singleElement.Stakers})
+		store.Set([]byte(singleElement.Key), bz)
+	}
+	return nil
+}
```
It follows the code style of Cosmos SDK, it's unnecessary to address the error here.

TimmyExogenous commented on x/delegation/types/genesis.go:
```diff
@@ -1,35 +1,71 @@
-// Validate performs basic genesis state validation returning an error upon any
-// failure.
-func (gs GenesisState) Validate() error {
+func ValidateIDAndOperator(stakerID, assetID, operator string) error {
+	// validate the operator address
+	if _, err := sdk.AccAddressFromBech32(operator); err != nil {
+		return xerrors.Errorf(
+			"invalid operator address for operator %s", operator,
+		)
+	}
+	_, stakerClientChainID, err := assetstypes.ValidateID(stakerID, true, false)
+	if err != nil {
+		return xerrors.Errorf(
+			"invalid stakerID: %s",
+			stakerID,
+		)
+	}
+	_, assetClientChainID, err := assetstypes.ValidateID(assetID, true, false)
+	if err != nil {
+		return xerrors.Errorf(
+			"invalid assetID: %s",
+			assetID,
+		)
+	}
+	if stakerClientChainID != assetClientChainID {
+		return xerrors.Errorf(
+			"the client chain layerZero IDs of the staker and asset are different, stakerID:%s, assetID:%s",
+			stakerID, assetID)
+	}
+	return nil
+}
```
Thanks, it will be updated in the next commit

TimmyExogenous commented on x/reward/keeper/claim_reward_test.go:
```diff
@@ -43,5 +43,5 @@ func (suite *RewardTestSuite) TestClaimWithdrawRequest() {
 	assetInfo, err := suite.App.AssetsKeeper.GetStakingAssetInfo(suite.Ctx, assetID)
 	suite.NoError(err)
-	suite.Equal(sdkmath.NewInt(10).Add(assets[assetID].StakingTotalAmount), assetInfo.StakingTotalAmount)
+	suite.Equal(sdkmath.NewInt(10).Add(assets[0].StakingTotalAmount), assetInfo.StakingTotalAmount)
```
This change is because the returned assets type has been changed to a list from a map. So it can't be changed as the suggestion.

TimmyExogenous commented on x/assets/keeper/operator_asset.go:
```diff
@@ -10,10 +10,46 @@ import (
-func (k Keeper) GetOperatorAssetInfos(ctx sdk.Context, operatorAddr sdk.Address, assetsFilter map[string]interface{}) (assetsInfo map[string]*assetstype.OperatorAssetInfo, err error) {
-	ret := make(map[string]*assetstype.OperatorAssetInfo, 0)
+// AllOperatorAssets
+func (k Keeper) AllOperatorAssets(ctx sdk.Context) (operatorAssets []assetstype.AssetsByOperator, err error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), assetstype.KeyPrefixOperatorAssetInfos)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]assetstype.AssetsByOperator, 0)
+	var previousOperator string
+	for ; iterator.Valid(); iterator.Next() {
+		keyList, err := assetstype.ParseJoinedStoreKey(iterator.Key(), 2)
+		if err != nil {
+			return nil, err
+		}
+		operator, assetID := keyList[0], keyList[1]
+		if previousOperator != operator {
+			assetsByOperator := assetstype.AssetsByOperator{
+				Operator:    operator,
+				AssetsState: make([]assetstype.AssetByID, 0),
+			}
+			ret = append(ret, assetsByOperator)
+		}
+		var assetInfo assetstype.OperatorAssetInfo
+		k.cdc.MustUnmarshal(iterator.Value(), &assetInfo)
+		index := len(ret) - 1
+		ret[index].AssetsState = append(ret[index].AssetsState, assetstype.AssetByID{
+			AssetID: assetID,
+			Info:    assetInfo,
+		})
+		previousOperator = operator
+	}
+	return ret, nil
+}
```
This implementation will iterate over the operator twice, so why suggest this?

TimmyExogenous commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -54,9 +55,9 @@
 message QueryAssetInfoResponse {
   // asset_infos is the response for the staker asset info, indexed by the asset id.
-  map<string, StakerAssetInfo> asset_infos = 1;
+  repeated DepositByAsset asset_infos = 1 [(gogoproto.nullable) = false];
```
This change is for future pagination when implementing the query RPC, and all impacts on the client side have been considered.

TimmyExogenous commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -74,9 +75,9 @@
 message QueryOperatorAssetInfosResponse {
   // asset_infos is the response for the operator asset info, indexed by the asset id.
-  map<string, OperatorAssetInfo> asset_infos = 1;
+  repeated AssetByID asset_infos = 1 [(gogoproto.nullable) = false];
```
This change is for future pagination when implementing the query RPC, and all impacts on the client side have been considered.

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -194,6 +192,33 @@ func (k *Keeper) GetOptedInAVSForOperator(ctx sdk.Context, operatorAddr string)
 	return avsList, nil
 }
 
+func (k *Keeper) SetAllOptedInfo(ctx sdk.Context, optedStates []operatortypes.OptedState) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), operatortypes.KeyPrefixOperatorOptedAVSInfo)
+	for i := range optedStates {
+		state := optedStates[i]
+		bz := k.cdc.MustMarshal(&state)
+		store.Set([]byte(state.Key), bz)
+	}
+	return nil
+}
```
This function is only used in the `InitGenesis`, and it will panic when an error occurs. So the suggestion is unnecessary.

TimmyExogenous commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -249,6 +276,33 @@ func (k *Keeper) GetStakersByOperator(ctx sdk.Context, operator, assetID string)
 	return stakerList, nil
 }
 
+func (k Keeper) AllStakerList(ctx sdk.Context) (stakerList []delegationtype.StakersByOperator, err error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), delegationtype.KeyPrefixStakersByOperator)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]delegationtype.StakersByOperator, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		var stakers delegationtype.StakerList
+		k.cdc.MustUnmarshal(iterator.Value(), &stakers)
+		ret = append(ret, delegationtype.StakersByOperator{
+			Key:     string(iterator.Key()),
+			Stakers: stakers.Stakers,
+		})
+	}
+	return ret, nil
+}
```
It follows the code style of Cosmos SDK. It's unnecessary to change here.

TimmyExogenous commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -14,6 +14,33 @@
 
 type DelegationOpFunc func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) error
 
+func (k Keeper) AllDelegationStates(ctx sdk.Context) (delegationStates []delegationtype.DelegationStates, err error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), delegationtype.KeyPrefixRestakerDelegationInfo)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]delegationtype.DelegationStates, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		var stateInfo delegationtype.DelegationAmounts
+		k.cdc.MustUnmarshal(iterator.Value(), &stateInfo)
+		ret = append(ret, delegationtype.DelegationStates{
+			Key:    string(iterator.Key()),
+			States: stateInfo,
+		})
+	}
+	return ret, nil
+}
```
It follows the code style of Cosmos SDK. It's unnecessary to change here.

TimmyExogenous commented on x/operator/keeper/consensus_keys.go:
```diff
@@ -438,3 +441,93 @@ func (k Keeper) ClearPreviousConsensusKeys(ctx sdk.Context, chainID string) {
+	return nil
+}
+
+func (k *Keeper) GetAllPrevConsKeys(ctx sdk.Context) ([]types.PrevConsKey, error) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{types.BytePrefixForOperatorAndChainIDToPrevConsKey})
+	defer iterator.Close()
+
+	ret := make([]types.PrevConsKey, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		var consKey tmprotocrypto.PublicKey
+		k.cdc.MustUnmarshal(iterator.Value(), &consKey)
+		ret = append(ret, types.PrevConsKey{
+			Key:          hexutil.Encode(iterator.Key()),
+			ConsensusKey: hexutil.Encode(consKey.GetEd25519()),
+		})
+	}
+	return ret, nil
+}
```
I will record this issue and add the related tests when I have some time.

TimmyExogenous commented on x/operator/keeper/consensus_keys.go:
```diff
@@ -438,3 +441,93 @@
+	return nil
+}
+
+func (k *Keeper) GetAllOperatorKeyRemovals(ctx sdk.Context) ([]types.OperatorKeyRemoval, error) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{types.BytePrefixForOperatorKeyRemovalForChainID})
+	defer iterator.Close()
+
+	ret := make([]types.OperatorKeyRemoval, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		ret = append(ret, types.OperatorKeyRemoval{
+			Key: hexutil.Encode(iterator.Key()),
+		})
+	}
+	return ret, nil
+}
```
I will record this issue and add the related tests when I have some time.

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous commented on x/assets/keeper/keeper.go:
```diff
@@ -47,13 +47,13 @@ type IAssets interface {
 	SetStakingAssetInfo(ctx sdk.Context, info *assetstype.StakingAssetInfo) (err error)
 	GetStakingAssetInfo(ctx sdk.Context, assetID string) (info *assetstype.StakingAssetInfo, err error)
-	GetAllStakingAssetsInfo(ctx sdk.Context) (allAssets map[string]*assetstype.StakingAssetInfo, err error)
+	GetAllStakingAssetsInfo(ctx sdk.Context) (allAssets []assetstype.StakingAssetInfo, err error)
```
All related clients of this interface has been updated.

TimmyExogenous commented on x/assets/keeper/keeper.go:
```diff
@@ -47,13 +47,13 @@
+	GetAllStakingAssetsInfo(ctx sdk.Context) (allAssets []assetstype.StakingAssetInfo, err error)
 
-	GetStakerAssetInfos(ctx sdk.Context, stakerID string) (assetsInfo map[string]*assetstype.StakerAssetInfo, err error)
+	GetStakerAssetInfos(ctx sdk.Context, stakerID string) (assetsInfo []assetstype.DepositByAsset, err error)
```
All related clients of this interface have been updated.

TimmyExogenous commented on x/assets/keeper/keeper.go:
```diff
@@ -47,13 +47,13 @@
 	UpdateStakerAssetState(ctx sdk.Context, stakerID string, assetID string, changeAmount assetstype.DeltaStakerSingleAsset) (err error)
 
-	GetOperatorAssetInfos(ctx sdk.Context, operatorAddr sdk.Address, assetsFilter map[string]interface{}) (assetsInfo map[string]*assetstype.OperatorAssetInfo, err error)
+	GetOperatorAssetInfos(ctx sdk.Context, operatorAddr sdk.Address, assetsFilter map[string]interface{}) (assetsInfo []assetstype.AssetByID, err error)
```
All related clients of this interface has been updated.

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

MaxMustermann2 commented on x/assets/types/genesis.go:
```diff
@@ -25,13 +34,344 @@ func NewGenesis(
+// CommonValidation is used to check for duplicates in the input list
+// and validate the input information simultaneously.
+// It might be used for validating most genesis states.
+func CommonValidation[T any, V constraints.Ordered, D any](
```
Love it! Maybe move it to `utils/utils.go` because it can be reused across modules?

TimmyExogenous replied:
Thanks, I will move it to `utils/utils.go` in the next commit.

MaxMustermann2 commented on x/delegation/types/genesis.go:
```diff
@@ -139,4 +175,209 @@ func (gs GenesisState) Validate() error {
+}
+
+func (gs GenesisState) ValidateUndelegations() error {
+	validationFunc := func(_ int, undelegaion UndelegationRecord) error {
```
Typo in `undelegation`

TimmyExogenous replied:
Thanks, will fix it in the next commit

MaxMustermann2 commented on x/operator/client/cli/tx.go:
```diff
@@ -73,10 +74,12 @@ func CmdRegisterOperator() *cobra.Command {
+	)
 	// ApproveAddr may be different from the sender's address.
 	f.String(
 		FlagApproveAddr, "", "The address which is used to approve the delegations made to "+
```
Is this used anywhere?

TimmyExogenous replied:
I removed the check that enforced the earning address to be the same as the operator address when setting operator states. So added a flag to allow operators to set a different earning address during registration. However, since the earning address isn't currently being used, I’m unsure if we should allow it to be different from the operator address.

MaxMustermann2 commented on x/operator/keeper/consensus_keys.go:
```diff
@@ -438,3 +443,113 @@ func (k Keeper) ClearPreviousConsensusKeys(ctx sdk.Context, chainID string) {
+		}
+		ret = append(ret, types.PrevConsKey{
+			Key:          string(assetstypes.GetJoinedStoreKey(chainID, operatorAddr.String())),
+			ConsensusKey: hexutil.Encode(consKey.GetEd25519()),
```
I have created an `operatortypes.WrappedConsKey` structure, which may be useful for this encoding and decoding, as well as conversion of the key into different formats.

TimmyExogenous replied:
Thanks, I will update it in the next commit.

TimmyExogenous commented on x/assets/types/genesis.go:
```diff
@@ -25,13 +34,344 @@ func NewGenesis(
+// CommonValidation is used to check for duplicates in the input list
+// and validate the input information simultaneously.
+// It might be used for validating most genesis states.
+func CommonValidation[T any, V constraints.Ordered, D any](
+	slice []T,
+	seenFieldValue func(T) (V, D),
+	validation func(int, T) error,
+) (map[V]D, error) {
+	seen := make(map[V]D)
+	for i := range slice {
+		v := slice[i]
+		field, value := seenFieldValue(v)
+		// check for no duplicated element
+		if _, ok := seen[field]; ok {
+			return nil, errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"duplicate element: %v",
+				field,
+			)
+		}
+		// perform the validation
+		if err := validation(i, v); err != nil {
+			return nil, err
+		}
+		seen[field] = value
+	}
+	return seen, nil
+}
```
I will update it in the next commit.

TimmyExogenous committed:
use panic to address the error when exporting genesis
fix the unit tests
fix the issue for genesis exporting launch
define the common delimiter for combined key,stakerID and assetID. And add the check for the chainID

bwhour commented on app/ethtest_helper.go:
```diff
@@ -204,31 +203,26 @@ func genesisStateWithValSet(codec codec.Codec, genesisState simapp.GenesisState,
+			OperatorAddress: operator.String(),
+			OperatorInfo: operatortypes.OperatorInfo{
+				EarningsAddr:     operator.String(),
+				OperatorMetaInfo: "operator1",
```
this this operator1 hardcode

TimmyExogenous replied:
This function hasn't been used, it's used for test preparation, so use `opeartor1` for simple meta info.

bwhour commented on x/assets/keeper/genesis.go:
```diff
@@ -40,21 +41,55 @@ func (k Keeper) InitGenesis(ctx sdk.Context, data *types.GenesisState) {
+func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 	res := types.GenesisState{}
-	// TODO
+	params, err := k.GetParams(ctx)
```
why not use a single-line **if** statement to handle errors, assigning the value to 'res' upon **successfully retrieving data** ?

TimmyExogenous replied:
Because it returns a pointer type. As for the other fields, I will change them to a single-line according to your suggestion.

bwhour commented on x/operator/keeper/genesis.go:
```diff
@@ -9,14 +10,96 @@ import (
-func (Keeper) ExportGenesis(sdk.Context) *types.GenesisState {
-	// TODO
-	return types.DefaultGenesis()
+func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
```
just record error, should we  assign data when it is successful without any error ?

TimmyExogenous replied:
I am unsure if an error should influence the export of other fields. I can adjust it to return a null if there is any error.

TimmyExogenous replied:
Use panic to address the error according to Max's advice.

bwhour commented on x/operator/keeper/consensus_keys.go:
```diff
@@ -438,3 +443,114 @@ func (k Keeper) ClearPreviousConsensusKeys(ctx sdk.Context, chainID string) {
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]types.OperatorConsKeyRecord, 0)
```
I observed that the memory allocation for similar **slices** is currently set to **zero**. While this approach functions correctly, it is recommended estimating a capacity for these allocations to avoid reallocation each time.

TimmyExogenous replied:
we don't know the capacity of these slices, and Golang will allocate a page for the slice, there won't be much of an issue.

bwhour commented on x/delegation/keeper/genesis.go:
```diff
@@ -66,12 +31,48 @@ func (k Keeper) InitGenesis(
+	if err != nil {
+		ctx.Logger().Error(errorsmod.Wrap(err, "failed to get all associations").Error())
+	}
+	res.Associations = associations
```
directly assign data even though it is failed.

TimmyExogenous replied:
I am unsure if an error should influence the export of other fields. I can adjust it to return a null if there is any error.

TimmyExogenous replied:
Use panic to address the error according to Max's advice.

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous commented on x/reward/keeper/claim_reward_test.go:
```diff
@@ -36,12 +36,12 @@ func (suite *RewardTestSuite) TestClaimWithdrawRequest() {
 	assetInfo, err := suite.App.AssetsKeeper.GetStakingAssetInfo(suite.Ctx, assetID)
 	suite.NoError(err)
-	suite.Equal(sdkmath.NewInt(10).Add(assets[assetID].StakingTotalAmount), assetInfo.StakingTotalAmount)
+	suite.Equal(sdkmath.NewInt(10).Add(assets[0].StakingTotalAmount), assetInfo.StakingTotalAmount)
```
There is only one registered asset  when testing, so it's okay to use the index 0.

TimmyExogenous commented on x/delegation/types/genesis.go:
```diff
@@ -140,3 +95,174 @@ func (gs GenesisState) Validate() error {
+		return errorsmod.Wrap(ErrInvalidGenesisData, err.Error())
+	}
+	return nil
+}
```
thanks, I will fix the typo in the next commit.

TimmyExogenous commented on x/delegation/keeper/genesis.go:
```diff
@@ -66,12 +31,44 @@ func (k Keeper) InitGenesis(
 			panic(errorsmod.Wrap(err, "failed to associate operator with staker"))
 		}
 	}
+
+	// init the state from the general exporting genesis file
+	err := k.SetAllDelegationStates(ctx, gs.DelegationStates)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to set all delegation states"))
+	}
+	err = k.SetAllStakerList(ctx, gs.StakersByOperator)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to set all staker list"))
+	}
+	err = k.SetUndelegationRecords(ctx, gs.Undelegations)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to set all undelegation records"))
+	}
```
Panic will provide detailed context for debugging.

TimmyExogenous commented on x/delegation/keeper/genesis.go:
```diff
@@ -66,12 +31,44 @@
-	genesis := delegationtype.DefaultGenesis()
-	// TODO
-	return genesis
+func (k Keeper) ExportGenesis(ctx sdk.Context) *delegationtype.GenesisState {
+	res := delegationtype.GenesisState{}
+	var err error
+	res.Associations, err = k.GetAllAssociations(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all associations").Error())
+	}
+
+	res.DelegationStates, err = k.AllDelegationStates(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all delegation states").Error())
+	}
+	res.StakersByOperator, err = k.AllStakerList(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all staker list").Error())
+	}
+
+	res.Undelegations, err = k.AllUndelegations(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all undelegations").Error())
+	}
+	return &res
```
panic will provide detailed context information for debugging.

TimmyExogenous commented on x/operator/keeper/genesis.go:
```diff
@@ -9,14 +10,90 @@ import (
+	var err error
+	res.Operators = k.AllOperators(ctx)
+
+	res.OperatorRecords, err = k.GetAllOperatorConsKeyRecords(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all consensus keys for operators").Error())
+	}
+
+	res.OptStates, err = k.GetAllOptedInfo(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all opted info").Error())
+	}
+
+	res.AVSUSDValues, err = k.GetAllAVSUSDValues(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all AVS USD values").Error())
+	}
+
+	res.OperatorUSDValues, err = k.GetAllOperatorUSDValues(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all operator USD values").Error())
+	}
+
+	res.SlashStates, err = k.GetAllSlashStates(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to set all slashing info").Error())
+	}
+
+	res.PreConsKeys, err = k.GetAllPrevConsKeys(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all previous consensus keys").Error())
+	}
+
+	res.OperatorKeyRemovals, err = k.GetAllOperatorKeyRemovals(ctx)
+	if err != nil {
+		panic(errorsmod.Wrap(err, "failed to get all key removals for operators").Error())
+	}
```
The node is offline when exporting genesis, so panic is fine.

TimmyExogenous commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -42,9 +43,9 @@
 message QueryAllStakingAssetsInfoResponse {
   // all_staking_assets_info is the response for all staking assets info.
-  map<string, StakingAssetInfo> all_staking_assets_info = 1;
+  repeated StakingAssetInfo all_staking_assets_info = 1 [(gogoproto.nullable) = false];
```
All client code has been considered and updated.

TimmyExogenous commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -74,9 +75,9 @@
 message QueryOperatorAssetInfosResponse {
   // asset_infos is the response for the operator asset info, indexed by the asset id.
-  map<string, OperatorAssetInfo> asset_infos = 1;
+  repeated AssetByID asset_infos = 1 [(gogoproto.nullable) = false];
```
All client code has been considered and updated.

TimmyExogenous commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -54,9 +55,9 @@
 message QueryAssetInfoResponse {
   // asset_infos is the response for the staker asset info, indexed by the asset id.
-  map<string, StakerAssetInfo> asset_infos = 1;
+  repeated DepositByAsset asset_infos = 1 [(gogoproto.nullable) = false];
```
All client code has been considered and updated.

TimmyExogenous commented:
I have addressed all code comments in this PR and tested the 3-nodes launching from Bootstrap and the exported genesis file. It can be merged when the review is finished.

bwhour commented on x/assets/types/genesis.go:
```diff
@@ -207,52 +181,142 @@ func (gs GenesisState) Validate() error {
+			// check that the asset is registered
+			// no need to check for the validity of the assetID, since
+			// an invalid assetID cannot be in the tokens map.
+			if _, ok := tokensTotalStaking[asset.AssetID]; !ok {
```
In assetValidationFunc, tokensTotalStaking[asset.AssetID] can be extracted to avoid redundant map access

> totalStaking, ok := tokensTotalStaking[asset.AssetID]
    if !ok {
...
if asset.Info.TotalAmount.Add(asset.Info.PendingUndelegationAmount).GT(totalStaking) {

TimmyExogenous replied:
Thanks, I will fix it in the next commit.

bwhour commented on x/operator/types/keys.go:
```diff
@@ -155,6 +209,31 @@ func KeyForOperatorKeyRemovalForChainID(addr sdk.AccAddress, chainID string) []b
 
+func ParseKeyForOperatorKeyRemoval(key []byte) (addr sdk.AccAddress, chainID string, err error) {
+	// Check if the key has at least 20 byte for the operator and eight byte for the chainID length
+	if len(key) < AccAddressLength+8 {
```
The  `AccAddressLength + 8` is recommended to be extracted as a variable or constant to enhance maintainability.

TimmyExogenous replied:
I will use a constant to replace 8 in the next commit.

bwhour reviewed (approved):
Added two another comments. 
The whole looks good.

TimmyExogenous committed:
add a constant ByteLengthForUint64 and fix the issue identified by code comment

trestinlsd commented:
The local_node script needs to be updated to address the issue of script startup: failed to unmarshal assets genesis state: unknown field "total_supply" in types.AssetInfo

TimmyExogenous committed:
fix the issue regarding starting a local node

TimmyExogenous commented:
> The local_node script needs to be updated to address the issue of script startup: failed to unmarshal assets genesis state: unknown field "total_supply" in types.AssetInfo

fixed

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous commented on x/dogfood/keeper/genesis.go:
```diff
@@ -127,10 +127,16 @@ func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 		pubKey, _ := val.ConsPubKey()
 		// #nosec G703 // already validated
 		convKey, _ := cryptocodec.ToTmPubKeyInterface(pubKey)
+		addr := sdk.GetConsAddress(pubKey)
+		found, operatorAddr := k.operatorKeeper.GetOperatorAddressForChainIDAndConsAddr(ctx, avstypes.ChainIDWithoutRevision(ctx.ChainID()), addr)
+		if !found {
+			return false
+		}
```
It seems like returning true will stop the iteration, and the iteration should stop when the operator isn't found. So I will change them according to your refactored version.

TimmyExogenous commented on x/dogfood/keeper/genesis.go:
```diff
@@ -127,10 +127,16 @@
 			types.GenesisValidator{
-				PublicKey: hexutil.Encode(convKey.Bytes()),
-				Power:     val.GetConsensusPower(sdk.DefaultPowerReduction),
+				PublicKey:       hexutil.Encode(convKey.Bytes()),
+				Power:           val.GetConsensusPower(sdk.DefaultPowerReduction),
+				OperatorAccAddr: operatorAddr.String(),
```
It seems unnecessary because the callback function doesn't involve much complex logic here.

TimmyExogenous commented on app/ethtest_helper.go:
```diff
@@ -200,31 +199,26 @@ func genesisStateWithValSet(codec codec.Codec, genesisState simapp.GenesisState,
-							Value: &delegationtypes.ValueField{
-								Amount: depositAmount,
-							},
-						},
-					},
-				},
+			Key: string(singleStateKey),
+			States: delegationtypes.DelegationAmounts{
+				WaitUndelegationAmount: math.NewInt(0),
+				UndelegatableShare:     math.LegacyNewDecFromBigInt(depositAmount.BigInt()),
 			},
 		},
```
Thanks, this function hasn't been used yet. It might be refactored in the future.

TimmyExogenous commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -91,8 +91,8 @@
   ];
-  // wait_unbonding_amount is the amount that is waiting for undelegation.
-  string wait_unbonding_amount = 3
+  // pending_undelegation_amount is the amount that is waiting for undelegation.
+  string pending_undelegation_amount = 3
```
The influence of this breaking change has been considered and addressed

TimmyExogenous commented on testutil/utils.go:
```diff
@@ -159,74 +204,114 @@ func (suite *BaseTestSuite) SetupWithGenesisValSet(genAccs []authtypes.GenesisAc
 				},
 			},
 		},
 		{
-			StakerID: stakerID2,
-			Delegations: []delegationtypes.DelegatedSingleAssetInfo{
+			OperatorAddress: operator2.String(),
+			Chains: []operatortypes.ChainDetails{
 				{
-					AssetID: assetID,
-					PerOperatorAmounts: []delegationtypes.KeyValue{
-						{
-							Key: operator2.String(),
-							Value: &delegationtypes.ValueField{
-								Amount: depositAmount2,
-							},
-						},
-					},
+					ChainID:      chainIDWithoutRevision,
+					ConsensusKey: pubKey2.ToHex(),
 				},
 			},
 		},
```
Thanks, for this function, we can refine it in the future.

TimmyExogenous commented on testutil/utils.go:
```diff
@@ -159,74 +204,114 @@
-			EarningsAddr:     operator1.String(),
-			OperatorMetaInfo: "operator1",
-			Commission:       stakingtypes.NewCommission(sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec()),
+			OperatorAddress: operator1.String(),
+			OperatorInfo: operatortypes.OperatorInfo{
+				EarningsAddr:     operator1.String(),
+				OperatorMetaInfo: "operator1",
+				Commission:       stakingtypes.NewCommission(sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec()),
+			},
 		},
 		{
-			EarningsAddr:     operator2.String(),
-			OperatorMetaInfo: "operator2",
-			Commission:       stakingtypes.NewCommission(sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec()),
+			OperatorAddress: operator2.String(),
+			OperatorInfo: operatortypes.OperatorInfo{
+				EarningsAddr:     operator2.String(),
+				OperatorMetaInfo: "operator2",
+				Commission:       stakingtypes.NewCommission(sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec()),
+			},
```
Thanks, for this function, we can refine it in the future.

TimmyExogenous commented on testutil/utils.go:
```diff
@@ -125,14 +129,55 @@
+	operatorAssets := []assetstypes.AssetsByOperator{
+		{
+			Operator: operator1.String(),
+			AssetsState: []assetstypes.AssetByID{
+				{
+					AssetID: assetID,
+					Info: assetstypes.OperatorAssetInfo{
+						TotalAmount:               depositAmount,
+						PendingUndelegationAmount: sdk.ZeroInt(),
+						TotalShare:                sdk.NewDecFromBigInt(depositAmount.BigInt()),
+						OperatorShare:             sdk.NewDecFromBigInt(depositAmount.BigInt()),
+					},
+				},
+			},
+		},
+		{
+			Operator: operator2.String(),
+			AssetsState: []assetstypes.AssetByID{
+				{
+					AssetID: assetID,
+					Info: assetstypes.OperatorAssetInfo{
+						TotalAmount:               depositAmount2,
+						PendingUndelegationAmount: sdk.ZeroInt(),
+						TotalShare:                sdk.NewDecFromBigInt(depositAmount2.BigInt()),
+						OperatorShare:             sdk.NewDecFromBigInt(depositAmount2.BigInt()),
 					},
 				},
 			},
 		},
 	}
+	assetsGenesis := assetstypes.NewGenesis(
+		assetstypes.DefaultParams(),
+		suite.ClientChains, []assetstypes.StakingAssetInfo{
+			{
+				AssetBasicInfo:     suite.Assets[0],
+				StakingTotalAmount: depositAmount.Add(depositAmount2),
+			},
+		}, depositsByStaker, operatorAssets,
+	)
+	genesisState[assetstypes.ModuleName] = app.AppCodec().MustMarshalJSON(assetsGenesis)
```
Thanks, for this function, we can refine it in the future.

TimmyExogenous commented on testutil/utils.go:
```diff
@@ -159,74 +204,114 @@
 			},
 		},
 	}
+	avsAddr := avstypes.GenerateAVSAddr(chainIDWithoutRevision)
+	optStates := []operatortypes.OptedState{
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(operator1.String(), avsAddr)),
+			OptInfo: operatortypes.OptedInfo{
+				OptedInHeight:  1,
+				OptedOutHeight: operatortypes.DefaultOptedOutHeight,
+			},
+		},
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(operator2.String(), avsAddr)),
+			OptInfo: operatortypes.OptedInfo{
+				OptedInHeight:  1,
+				OptedOutHeight: operatortypes.DefaultOptedOutHeight,
+			},
+		},
+	}
+	operatorUSDValues := []operatortypes.OperatorUSDValue{
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(avsAddr, operator1.String())),
+			OptedUSDValue: operatortypes.OperatorOptedUSDValue{
+				SelfUSDValue:   usdValue,
+				TotalUSDValue:  usdValue,
+				ActiveUSDValue: usdValue,
+			},
+		},
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(avsAddr, operator2.String())),
+			OptedUSDValue: operatortypes.OperatorOptedUSDValue{
+				SelfUSDValue:   usdValue2,
+				TotalUSDValue:  usdValue2,
+				ActiveUSDValue: usdValue2,
+			},
+		},
+	}
```
Thanks, for this function, we can refine it in the future.

TimmyExogenous commented on testutil/utils.go:
```diff
@@ -237,7 +322,21 @@
 		},
 	}
-	delegationGenesis := delegationtypes.NewGenesis(delegationsByStaker, associations)
+	stakersByOperator := []delegationtypes.StakersByOperator{
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(operator1.String(), assetID)),
+			Stakers: []string{
+				stakerID1,
+			},
+		},
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(operator2.String(), assetID)),
+			Stakers: []string{
+				stakerID2,
+			},
+		},
+	}
```
Thanks, for this function, we can refine it in the future.

TimmyExogenous commented on testutil/utils.go:
```diff
@@ -159,74 +204,114 @@
+	genesisState[operatortypes.ModuleName] = app.AppCodec().MustMarshalJSON(operatorGenesis)
+
+	// x/delegation
+	delegationStates := []delegationtypes.DelegationStates{
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(stakerID1, assetID, operator1.String())),
+			States: delegationtypes.DelegationAmounts{
+				WaitUndelegationAmount: math.NewInt(0),
+				UndelegatableShare:     math.LegacyNewDecFromBigInt(depositAmount.BigInt()),
+			},
+		},
+		{
+			Key: string(assetstypes.GetJoinedStoreKey(stakerID2, assetID, operator2.String())),
+			States: delegationtypes.DelegationAmounts{
+				WaitUndelegationAmount: math.NewInt(0),
+				UndelegatableShare:     math.LegacyNewDecFromBigInt(depositAmount2.BigInt()),
+			},
+		},
+	}
```
Thanks, for this function, we can refine it in the future.

TimmyExogenous commented on x/operator/types/genesis.go:
```diff
@@ -57,27 +84,466 @@ func (gs GenesisState) Validate() error {
+				"ValidatePrevConsKeys can't parse the combined key, %+v",
+				prevConsKey,
+			)
+		}
+
+		chainID, operator := keys[0], keys[1]
+		if !utils.IsValidChainIDWithoutRevision(chainID) {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"ValidatePrevConsKeys: invalid chainID without revision, operator %s: chainID: %s", operator, chainID,
+			)
+		}
+		// check that the operator is registered
+		if _, ok := operators[operator]; !ok {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"unknown operator address for the previous consensus key, %+v",
+				prevConsKey,
+			)
+		}
+		if wrappedKey := NewWrappedConsKeyFromHex(
+			prevConsKey.ConsensusKey,
+		); wrappedKey == nil {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"invalid previous consensus key for operator, %+v", prevConsKey,
+			)
+		}
+		// todo: not sure if the duplication of previous consensus keys needs to be checked
+		return nil
+	}
+	seenFieldValueFunc := func(prevConsKey PrevConsKey) (string, struct{}) {
+		return prevConsKey.Key, struct{}{}
+	}
+	_, err := utils.CommonValidation(gs.PreConsKeys, seenFieldValueFunc, validationFunc)
+	if err != nil {
+		return errorsmod.Wrap(ErrInvalidGenesisData, err.Error())
+	}
+	return nil
+}
```
Thanks, the unit test needs to be added in the future.

TimmyExogenous commented on x/operator/types/genesis.go:
```diff
@@ -57,27 +84,466 @@
+	return avs, nil
+}
+
+func (gs GenesisState) ValidateAVSUSDValues(optedAVS map[string]struct{}) (map[string]DecValueField, error) {
+	avsUSDValueMap := make(map[string]DecValueField, 0)
+	validationFunc := func(_ int, avsUSDValue AVSUSDValue) error {
+		if !common.IsHexAddress(avsUSDValue.AVSAddr) {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"the AVS address isn't an ethereum hex address, avsUSDValue: %+v",
+				avsUSDValue,
+			)
+		}
+		if _, ok := optedAVS[avsUSDValue.AVSAddr]; !ok {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"the avs address should be in the opted-in map, avsUSDValue: %+v", avsUSDValue,
+			)
+		}
+		if avsUSDValue.Value.Amount.IsNil() ||
+			avsUSDValue.Value.Amount.IsNegative() {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"avsUSDValue is nil or negative, avsUSDValue: %+v", avsUSDValue,
+			)
+		}
+		avsUSDValueMap[avsUSDValue.AVSAddr] = avsUSDValue.Value
+		return nil
+	}
+	seenFieldValueFunc := func(usdValue AVSUSDValue) (string, struct{}) {
+		return usdValue.AVSAddr, struct{}{}
+	}
+	_, err := utils.CommonValidation(gs.AVSUSDValues, seenFieldValueFunc, validationFunc)
+	if err != nil {
+		return nil, errorsmod.Wrap(ErrInvalidGenesisData, err.Error())
+	}
+	return avsUSDValueMap, nil
+}
```
Thanks, the unit test needs to be added in the future.

TimmyExogenous commented on x/operator/types/genesis.go:
```diff
@@ -2,53 +2,80 @@
 
 func NewGenesisState(
-	operators []OperatorInfo,
+	operators []OperatorDetail,
+	operatorConsKeys []OperatorConsKeyRecord,
+	optStates []OptedState,
+	operatorUSDValues []OperatorUSDValue,
+	avsUSDValues []AVSUSDValue,
+	slashStates []OperatorSlashState,
+	prevConsKeys []PrevConsKey,
+	operatorKeyRemovals []OperatorKeyRemoval,
```
Thanks, the unit test needs to be added in the future.

TimmyExogenous committed:
stop the iteration if the operator address isn't found when exporting genesis state for dogfood

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
rebase and correct the invalid log printing

TimmyExogenous commented on x/delegation/types/genesis.go:
```diff
@@ -140,3 +95,174 @@ func (gs GenesisState) Validate() error {
+				)
+			}
+			if stakerClientChainID != assetClientChainID {
+				return errorsmod.Wrapf(ErrInvalidGenesisData, "the client chain layerZero IDs of the staker and asset are different,key:%s stakerID:%s", stakersByOperator.Key, stakerID)
```
I will update it in the other PR.

TimmyExogenous commented on testutil/utils.go:
```diff
@@ -105,6 +107,8 @@ func (suite *BaseTestSuite) SetupWithGenesisValSet(genAccs []authtypes.GenesisAc
 	suite.Powers = []int64{power, power2}
 	depositAmount := math.NewIntWithDecimal(power, 6)
 	depositAmount2 := math.NewIntWithDecimal(power2, 6)
+	usdValue := math.LegacyNewDec(power)
+	usdValue2 := math.LegacyNewDec(power2)
```
There isn't a function named `NewDec` in the package `cosmossdk.io/math`
